lib/ecto/type.ex

```diff
@@ -295,6 +295,10 @@ defmodule Ecto.Type do
     dump_array(type, value, [], false)
   end
 
+  def dump({:map, model}, %{__struct__: model} = struct) do
+    {:ok, struct |> Map.from_struct |> Map.delete(:__meta__)}
```

Comment:
You will need to add a clause to the `of_base_type?` function in this module. Check the current map one for examples. Don't forget to add unit tests for the changes in this module too. Use the `:map` type commit for reference if you want to: 6f6b300273d352a326c040f66b7c849ace7e5474